### ISSUE TITLE
[API 8+] Add (back) timings support 

### ISSUE THREAD

**Issue by Rasgnarok** [SpongePowered/Sponge#3649]
### Is this request specific to one implementation of Sponge?

No

### Sponge Version

8.0.0

### What are you requesting?

As we all know, timings has been slated for removal for API 8, after much internal and external discussion. Two days ago, it was removed (https://github.com/SpongePowered/Sponge/commit/235f3dd293e401834a5d01f862ed2ccb43239650), completing one more milestone towards API 8's completion. 

While I understand why timings was a _lot_ to maintain when it comes to continuous support, timings also provided additional context on performance no other system provides, from WarmRoast to Spark. To diagnose performance in feature parity to timings, you would need to track over X minutes a lot more than simply the main thread. Timings' strength stemmed from the contextual timeline display of fluctuating variables such as playercount, chunks, entity count, TPS, etc. 

Therefore, I am suggesting the re-adding of timings to the Sponge API platform, either included by default within it, or as an external mixin plugin for diagnosing purposes. After all, platforms like Bukkit have continuously supported timings even with resources like WarmRoast or Spark around. In fact, even Spark itself is meant to be used within the context of other statistical data... including timings (https://github.com/lucko/spark/issues/5#issuecomment-405154388).

**Comment by vaniron** [comment]
I (as owner of multiple servers) am also in for re-adding of timings to sponge 8. Its helpfull and i would like to see this feature in new versions of sponge.

**Comment by witchyface** [comment]
Please re-add timings. It's an extremely important resource for server owners and admins.

**Comment by HackuJacku** [comment]
Thanks for adding this Ras, timings one of the essential resources for understanding performance-related issues. The removal of timings in API 8 would be a huge loss to all server owners/admins

**Comment by johnfriedrich** [comment]
Since the early days of timings 2013, or at least I remember using it since then, timings were an easy and quick command to check a servers rough health status. Timings got better and better and with v2 it actually gave admins a very good overview how the server was doing in the last 30 minutes. Entity/Chunk/tps over etc. time. 

Imagine having 100 people online and you are offering a special event for players or are testing a new plugin that could degrade performance, the first time on live servers. You can test plugin performance with a few test players, but there wil always be the first time with many players.
So what tool do you wanna use to check if performance is great all the time during it? Yes, you can use /tps. But what if tps actually reported bad in the last 5 minutes? 
You cannot see why.

So you may say "use Spark/warmroast". Yes, you could use Spark to get a snapshot, but there is that entry barrier. You need spark installed, plugin commands are a bit more complex than timings ones and its not that straight forward to read.

Back to my example. Without timings there is no way to check for admins if there was something going nuts during the last 30 minutes during that event. You would only notice lagspikes or you could spam /tps all the time. But you wont see the reason for lag.
In case staff members are noticing lag spikes they need to call an admin that take Spark timings RIGHT now to identify the issue.
Far more simple if you get some time when using timings because its always recorded.

You also might be saying that timings are inaccurate (Sponge impl). Yes thats true. BUT: There is no entry barrier. As a newbie owner you can simple run /timings, post the link and other can point you in the right direction where the lag is coming come.
And yes, in some cases you need a Spark or warmroast report to exactly define where the was coming from.

In short, timings give you a perfect, snappy overview, be it tps/events/configs/playercount to tps/chunkcount/entitycount relation and MUCH more, for a specific moment and, thats the most important, it gives you that view retrospectively over the last 30 minutes.
Also the output is far more easier to read. I am saying that as someone that used warmroast, timings and other profilers since day one back in 2013 and is comfortable with reading and understanding them.
So please, bring it back, this tool is heavily needed

**Comment by Rasgnarok** [comment]
> You also might be saying that timings are inaccurate (Sponge impl). Yes thats true. BUT: There is no entry barrier. As a newbie owner you can simple run /timings, post the link and other can point you in the right direction where the lag is coming come.

I think there is a lot to say about the early implementations of timings, where it not only was accurate, but also was heavily favoured by mod and plugin devs alike. Having an in-house profiling solution the standard for good reason - a well implemented timings provides the perfect balance between performance history and lightweight accessibility. Once can't say the same of Spark, or WarmRoast.

**Comment by PGGreed** [comment]
Timings are an essential tool to many servers, it helped me identify key problems with my server with ease and was quick and reliable. Do bring back timings cause otherwise that's a bad move on Sponge's part 👎

**Comment by Mikirae** [comment]
The removal of timings creates an unnecessary entry barrier for those looking to figure out lag issues within their server.
Many servers do not have the capability or knowledge in order to read things which are as complex as Spark reports or the like, but the vast majority of owners can quite easily deduce the cause of their lag from a well-written timings report.

**Comment by dualspiral** [comment]
We're not re-adding timings in the way they were added. A comment above sums this up well:

> You also might be saying that timings are inaccurate (Sponge impl). Yes thats true. 

**Bad data is worse than no data.** If the timings are inaccurate, then they should not be there - and that was a factor in our decisions.

We did think about the wider implications of the removal of timings when we made this decision. 

* Accuracy of the data - it wasn't that accurate.
* Maintenance cost on our end. We were trying to support an old system (Paper, which also uses the timing system, pretty much has an entirely different implementation - the updates were never ported to Sponge), and the implementation was old and not fit for purpose. We have our hands full sorting out Forge support, which is an entirely different beast in 1.16, worrying about timings is the least of our concerns.
* The timings views actually caused some confusion. The graphs at the top might have been nice, but people constantly mis-reading the lag views caused many a mis-diagnosis. For example, Nucleus got blamed _a lot_ for lag because it had a large tick time for /rtp, but it actually it was one tick, for perhaps twice the tick time, and not a continuous thing. We were rarely able to pinpoint a lag issue by looking at timings. Whether that was a problem in data analysis, or how we sent up the data, I don't know. 
* There are other plugins and methods for timings. Indeed, once we can get SF to 1.18.x, [there is the /jfr command that uses inbuilt Java systems to measure stuff](https://minecraft.fandom.com/wiki/Commands/jfr). Indeed, I'd be open to trying to enhance timings though that.

I know that the removal of this command was never going to be universally praised, but nor should we keep a system that is fundamentally flawed in our codebase, _especially_ if we have many other priorities.

That said, this is open source! I know many of you here are not developers, but you likely know people who are, have a bit of time, and might be interested in taking a swing at this. I am not opposed to seeing PRs that update the timings systems to be more modern, perhaps aligned more to what Paper does. Alternatively, if someone wants to create an external mixin plugin that re-introduces the code, by all means.

I'm not against a timings system being re-introduced to the main code base, but it needs to be re-written completely.

**Comment by Yeregorix** [comment]
I've been doing support on the Sponge discord for a few years now and I've seen a lot of Timings reports but I've rarely been able to extract anything other than "you have wrong GC flags" or "you should upgrade hardware". Most of the time, the reports were useless or misleading. I've been integrating Timings in my plugins, thinking this would help diagnosing performance but I ended up using Spark.

In my opinion, Spark and JFR are currently way better solutions on Sponge. Timings is an excellent idea, but the current integration with Sponge is very flawed. If someone really wants to dedicate its time to improve Timings on Sponge, feel free, but I doubt Sponge core devs have the time to.

Whatever is the best solution (Timings vs Spark vs future plugins), I don't see why Sponge should favor one over another. Timings should be an external plugin, like all other performance diagnostic tools.

**Comment by Niiohontesha** [comment]
Please re-add timings support, it would save a lot of people a lot of time and efficiency.

**Comment by dualspiral** [comment]
Please do not continue to comment "please re-add timings support", because that does not add anything to this thread. Read what I said, and if you have something constructive to contribute based on that _then_, please do so. Simply asking to re-add timings without addressing my comments adds nothing and will do nothing, and is quite frankly rude when I've spent time outlining our position and what we feel needs doing.

I haven't closed this issue because I very much believe it's in our interest to listen and I understand why some people feel such a command is useful, but it's also very much in our interest that everyone understands our position - especially in that timings were very flawed and somewhat useless in the context of 1.16.

I reiterate - Sponge is an open source project and if any developers want to take up the challenge of re-writing Timings to be a lot more accurate and performant, you can! If you want the current implementation, we'll not be reintegrating it, but someone is free to turn it into an external mixin plugin if they so wish.

**Comment by Hiroku** [comment]
I've had a similar experience trying to diagnose issues using timings, and have gradually learned that it's better to skip the timings step entirely and get straight to the Spark report. It was a wise move to drop it from the API; bad information is much worse than no information. And it's worth stressing, Spark is a spectacular alternative and should be the recommended way to diagnose performance issues until things move to 1.18. That it's harder than timings is like saying that a bicycle is easier than a car so you should continue using it for cross-country travel.

**Comment by johnfriedrich** [comment]
> **Bad data is worse than no data.** If the timings are inaccurate, then they should not be there - and that was a factor in our decisions.

That statement is true for the real world or enterprise applications. The data that was generated was not always inaccurate and even if it was inaccurate,it gave you a direction, which was very important. Its nice to know where to start. Here again, low entry barrier. Basic diagnosis for people that have Spark not (yet) installed.


> * Maintenance cost on our end. We were trying to support an old system (Paper, which also uses the timing system, pretty much has an entirely different implementation - the updates were never ported to Sponge), and the implementation was old and not fit for purpose. We have our hands full sorting out Forge support, which is an entirely different beast in 1.16, worrying about timings is the least of our concerns.

Cant say that much here, but when skimming through SpongeForge/Common 1.12.2 commits, there wasnt that much timings related.


> * The timings views actually caused some confusion. The graphs at the top might have been nice, but people constantly mis-reading the lag views caused many a mis-diagnosis. For example, Nucleus got blamed _a lot_ for lag because it had a large tick time for /rtp, but it actually it was one tick, for perhaps twice the tick time, and not a continuous thing. We were rarely able to pinpoint a lag issue by looking at timings. Whether that was a problem in data analysis, or how we sent up the data, I don't know.

In that example indeed it was not nucleus itself, but maybe slow chunk loading or disk, but still, admins could see where it was coming from roughly.

> * The timings views actually caused some confusion. The graphs at the top might have been nice, but people constantly mis-reading the lag views caused many a mis-diagnosis

Well, when reading help discords, spark view and warmroast view is a lot harder to understand. mis-diagnosis is very often also a thing with spark.



> * There are other plugins and methods for timings. Indeed, once we can get SF to 1.18.x, [there is the /jfr command that uses inbuilt Java systems to measure stuff](https://minecraft.fandom.com/wiki/Commands/jfr). Indeed, I'd be open to trying to enhance timings though that.

So far so good, but we are talking (mainly) about 1.16

Its a bit sad that developers are saying this timings thing is useless compared to maintenance cost and therefore remove it.
In the end, we, admins are using this software on small/big servers with 20 or 1000 people. When admins are saying they find it useful, this word is definitely more worth than a developer saying "its useless and delivers inaccurate data, lets remove it"

You as developer probably did not even find the chance to use that feature on a regular basis on servers with many players online where its vital to track entitycount over the time, playercount over the time. How the server was doing over the time with settings X changed etc.

And yes, I have seen many reports in Sponge Discord where admins linked reports and falsely blamed Nulcues/Sponge for something where it wasnt Nucleus/Sponge. Sometimes because they didn't know better or the report was wrong. But besides that, there are probably hundred or even thousand admins that are using this report on a daily basis(without posting it somewhere) that do need that and do find it useful.

That feature was also helpful to easily compare servers health over a timeframe when you have changed settings etc.
Admins do find this very useful and at least for me and even if no one will care, it renders Sponge 1.16 not viable for me.

This whole removal thing felt a bit like when a Sponge dev would remove the stop command with the reason that he uses the "X" on the powershell console all the time while developing anyway and the stop command fails from time to time with bad plugins.

**Comment by dualspiral** [comment]
> That statement is true for the real world or enterprise applications.

This isn't the real world? It feels pretty real when we get people coming for support with timings that tell us nothing, or worse, look bad but actually were a one tick blip on some command. For some, gaming is part of their real world, a coping mechanism for all the bad out there, supports those with anxiety or autistic tendencies. Part of why I got into Minecraft modding was because of family friends who could use Minecraft as a way to help their children, and I could build little things that helped their children cope with life.

So don't tell me this isn't real world. For some, it really is.

Bad data has also led to our support getting requests with no context, just a timings report that says nothing. That's our support's time and effort. It also wastes the time of admins who read it and try to fix things, only to find their problem wasn't a problem.

> The data that was generated was not always inaccurate and even if it was inaccurate,it gave you a direction, which was very important.

Not always, as I said, it led to mis-diagnosis because of bad data. Some of that is down to the implementation doing the wrong thing.

> In that example indeed it was not nucleus itself, but maybe slow chunk loading or disk, but still, admins could see where it was coming from roughly.

My point was that it wasn't though. Admins were looking at one tick taking a long time, rather than multiple ticks just lagging a little. In that particular scenario, their entire reasoning was wrong - Nucleus was top of the list but it was a half second pause, not continuous 70ms ticks.

That's more a problem with how timings were displayed, but that was my point which you missed by telling me the cause of the single tick issue, not the underlying "my server is continuously lagging" problem.

> Its a bit sad that developers are saying this timings thing is useless compared to maintenance cost and therefore remove it.
> In the end, we, admins are using this software on small/big servers with 20 or 1000 people. When admins are saying they find it useful, this word is definitely more worth than a developer saying "its useless and delivers inaccurate data, lets remove it"

In the end, we developers are the ones who are maintaining the software, and we made that judgement call as no-one else has been pitching in. As I said, Sponge is open source, many server admins have developers on their rosters, why not contribute back to the project and help build a better timings, at least one that is on par with Paper?

> You as developer probably did not even find the chance to use that feature on a regular basis on servers with many players online where its vital to track entitycount over the time, playercount over the time. How the server was doing over the time with settings X changed etc.

Don't ever assume. Some of us run servers and do eat our own dogfood.

> This whole removal thing felt a bit like when a Sponge dev would remove the stop command with the reason that he uses the "X" on the powershell console all the time while developing anyway and the stop command fails from time to time with bad plugins.

How is that even a comparison? The removal of timings was to do with the implementation and how it was bad data, a fundamental issue, not a workflow one.

---

This is what I find quite extraordinary about some of these statements - we're getting significant criticism about this decision and while I've made the position clear, I'm still being told "this is bad, Sponge should have this, etc. etc."., yet we're getting no-one coming up to the plate to help us with a better version of Timings. Indeed:

> Admins do find this very useful and at least for me and even if no one will care, it renders Sponge 1.16 not viable for me.

Comments like this aren't going to make us change our opinions, demands and what feels like blackmail aren't going to get anywhere. Please remember that:

* Sponge is a free product, we have never, and will never charge for it.
* We are volunteers who try to work on it in our spare time because we _want_ to make the best API we can. None of us get paid to do this, none of us want to get paid to do this.
* We are open source and we accept contributions, as I said before.

I have explained what the options are in a previous comment. I haven't closed this issue because I understand people's arguments, but equally, understand that we're not reintroducing what we had.

**Comment by Hiroku** [comment]
> That statement is true for the real world or enterprise applications. The data that was generated was not always inaccurate and even if it was inaccurate,it gave you a direction, which was very important. Its nice to know where to start. 
>Here again, low entry barrier. Basic diagnosis for people that have Spark not (yet) installed.
It's not nice to know where to start when you can't have any confidence in it being correct. People were on dual's case for Nucleus in timings, so they must have felt very nice knowing where to start - but they were also wrong.

This is the crucial problem with timings, in my opinion: Some server owners think it works because once in a blue moon it turns up with a result that vaguely points in the right direction. The code for it is outdated, painful to maintain unless I miss my guess, and generates an avalanche of incorrect diagnoses - but some server owners think it works, have been conditioned to think it works by its accuracy in the past, so they fight tooth and nail to protect it. They don't have the development knowledge to know how misleading it is, how much damage it causes to be misled, or even the kind of time that needs to be spent (free of charge!) to maintain the same tool that's generating so many misdiagnoses. What they do know is that it makes them _feel_ like they have a grip on what's going on, regardless of how untrue that is, and they know that they don't want to learn how to generate or read Spark reports for real answers.

The TL;DR is this: Developers working for free should not spend their time working on only seasonally useful tools for the _illusion_ of benefit to server owners that simply don't want to learn how to use a working one. As dualspiral has said repeatedly, if you're so keen on this half-measure placebo, put some money into having it maintained.

**Comment by Rasgnarok** [comment]
> As dualspiral has said repeatedly, if you're so keen on this half-measure placebo, put some money into having it maintained.

This discussion has strayed far, far from my original post. I am not asking for anyone's opinion on a hypothetical server spending, I am asking to consider the other side of the argument. There was no need for this adversity, and it has no place here.

> In the end, we developers are the ones who are maintaining the software, and we made that judgement call as no-one else has been pitching in. 

I didn't realize this had been a public consideration before removal, but if it was, it likely wasn't made public enough - I for one would've pitched in sooner. I understand why it was an internal decision, someone has to pull the trigger eventually on big changes, and this is one of them.

> Not always, as I said, it led to mis-diagnosis because of bad data. Some of that is down to the implementation doing the wrong thing.

Absolutely, which is why I pointed to an implementation issue being at fault for timings' lack of reliability. I remember 1.10.2 as an entirely different beast to 1.12.2 when it came to performance diagnosis, and timings was used widely then.

---- 
> I have explained what the options are in a previous comment. I haven't closed this issue because I understand people's arguments, but equally, understand that we're not reintroducing what we had.

This brings me back to my main point, which yes, was the use-case of timings concerning accessibility, but also the standard in place vis-a-vis other profilers. I've mentioned this to Luck prior - when Spark became the norm due to lack of Timings' maintaining. His answer made sense, Spark doesn't need to relay as much information as timings because it is meant to be used in conjunction, not instead of.

> Therefore, I am suggesting the re-adding of timings to the Sponge API platform, either included by default within it, or as an external mixin plugin for diagnosing purposes.

The first option of my suggestion is a no-go, but is there the possibility of the second one?

**Comment by dualspiral** [comment]
Someone external can do it, it won't be the core team.
